### ISSUE TITLE
Unpredictable Behavior when Upperdir is also Mountpoint

### ISSUE THREAD

**Issue by Swivelgames** [containers/fuse-overlayfs#393]
There's a use-case where the `upperdir` provided is also the `mountpoint`. In this scenario, the directory is fixed and utilized by a particular application or process, and already contains core files. Using OverlayFS, configuration files can be put in place and applied on top of this directory. However, while I'm not entirely sure if this use-case is entirely supported, it is also not forbidden. I would argue for one or the other, in order to prevent confusing or major errors.

Currently, when the `Upperdir` is also the `mountpoint`, there is quite a bit of unpredictable behavior that happens. Listing the mountpoint gives you different results depending on _what_ is listing the directory (for instance, `/bin/ls` gives a certain list, and other applications give a different listing). Still files that _are_ technically there can still be read and written, even by the applications that can't find them when listing a directory.

I'm fully aware of the _reason_ for this instability, because the same location is technically occupied by two different trees. In the worst case scenario, where the directory tree of the Upperdir and the ephemeral Mountpoint overlap, utilities can often times reference one _and_ the other, causing application errors and even, in the worst case scenarios that I've experienced, system instability and crashing.

<details>
<summary>Preserved from original post</summary>
<blockquote>

I would advocate for one of two options:

1) Ideally, an unknown level of effort to fix the stability issues and support the Upperdir and Mountpoint being the same location
2) ~Barring this type of scenario altogether~

~While I would obviously prefer number 1 here, I could certainly see the argument that this type of unpredictable behavior may not be easily wrangled, so I would advocate for number 2. Had I received an error in the outset saying `Cannot mount to upperdir`, `Upperdir cannot be the same as mountpoint`, or something similar, I probably would have gone with a different solution (albeit, far less elegant, like just simply symlinking, rsync, or something else).~

~Additionally, there should also probably some mechanism in place to bar the `mountpoint` from being a `lowerdir`.~

</blockquote>
</details>

After perusing different places, it seems that while `fuse-overlayfs` may or may not openly support it, I've found a number of different places where the `mountpoint` is set to either a `lowerdir` or the `upperdir`, which may make barring this behavior problematic for those doing this already. This would definitely break backwards compatibility since this is technically working with some degree of stability for others.

It might make sense to dig into this a bit deeper, instead, to figure out how it might be possible to remedy some of the stability issues.

Thoughts?

All in all, thanks for the awesome work that's already been put into this project. It's been fantastic for most all other uses I've had.

**Comment by Swivelgames** [comment]
In general, actually, I don't think this really has anything to do with any of the directories being the same, but probably more specifically, the `mountpoint` being not empty.

I think the `upperdir` being the same as the `mountpoint` might have simply uncovered most of the odd behavior.

**Comment by winnscode** [comment]
I think maybe the application that can't find the file still hold the old fd of the upperdir/mountpoint? IMHO it should be ok for overlayfs (both overlayfs-in-kernel and fuse-overlayfs) to use the upperdir or lowdir or any other non-empty directory as the mountpoint...

Before mounting, we'll get the origin fd of the upperdir when we open the upperdir by path. And after mounting, we would follow/traverse the mount and get the root of overlayfs intead when we open the same path. They are totally different objects in kernel (file/dentry/inode/fs), even though they are opened through the same path/name...

The overlayfs will open directoris of all layers and hold their fd for further access, like writting file to upper-layer, but the user application should drop the origin fd and switch to the new upperdir, then access the newly mounted file system.

**Comment by Swivelgames** [comment]
@giuseppe Any idea what could be going on here? This is starting to turn into a blocker, and my head is hurting tracking down the source of the issue :disappointed:

**Comment by giuseppe** [comment]
this use case is not supported.

fuse-overlayfs will be confused to lookup files on its own mountpoint, as it gets into loops when it uses paths.

**Comment by czheo** [comment]
@giuseppe does this statement also apply to `lowerdir`? having lowerdir as mountpoint is also not supported?

**Comment by giuseppe** [comment]
yes, the lowerdirs are supposed to be static (i.e. their content never changes), the kernel implementation has the same requirement

**Comment by Swivelgames** [comment]
@giuseppe While I'd love to argue for supporting this use-case, the main focus of this issue was:

- Either support it
- Or error on it

In my particular use-case, this is not a supported use-case, but it is allowed, and results in corruption of the files. If it's not a supported use-case, then **it should throw an error when attempted to prevent file corruption** 👍

This should be reopened and addressed. Recent activity suggests it isn't unheard of for users to attempt this.

**Comment by giuseppe** [comment]
I've reopened the issue.  If anyone is interested to open a PR to add a warning/error, please do so

**Comment by czheo** [comment]
> I've reopened the issue.  If anyone is interested to open a PR to add a warning/error, please do so

Need bump the major version if we do so since that's a backward incompatible change

**Comment by giuseppe** [comment]
no, it is not an incompatible change.  The current setup is not supported, it must not be used

**Comment by czheo** [comment]
> no, it is not an incompatible change. The current setup is not supported, it must not be used

There are existing users who have the same dir both in lower/upper dir and mountpoint. it's not failing, and in fact working okay today even though you claim that it's not supported. If you start to fail in a sudden, it's obviously an incompatible change from user's perspective no matter how you claim it.

**Comment by giuseppe** [comment]
it can't work correctly, it is just by coincidence.  You are messing with the underlying cache.

**Comment by czheo** [comment]
@giuseppe I believe you may have better context than me to make a correct call here. Would you please kindly shed some lights on how fuse-overlayfs would work incorrectly or "by coincidence" in this particular case `fuse-overlayfs -o lowerdir=a_dir:b_dir b_dir`? (It seems working okay as far as I can tell)

```bash
$ fuse-overlayfs --version
fuse-overlayfs: version 1.10
FUSE library version 3.14.0
using FUSE kernel interface version 7.31
fusermount3 version: 3.14.0

# test set up with 2 dirs
$ tree
.
├── a_dir
│   ├── a.txt
│   └── c.txt
└── b_dir
    ├── a.txt
    └── b.txt

# content of files
$ cat *_dir/*
a.txt in a_dir
c.txt in a_dir
a.txt in b_dir
b.txt in b_dir

# mountpoint b_dir is in lowerdir
$ fuse-overlayfs -o lowerdir=a_dir:b_dir b_dir
$ tree
.
├── a_dir
│   ├── a.txt
│   └── c.txt
└── b_dir
    ├── a.txt
    ├── b.txt
    └── c.txt

# content of b_dir is merged with a_dir as expected
$ cat b_dir/*
a.txt in a_dir
b.txt in b_dir
c.txt in a_dir
```

**Comment by giuseppe** [comment]
because fuse-overlayfs doesn't reload the directory content for every operation, so if you remove a file in the merged directory, it could still show up because the lower dir was not cleaned up.

Now it might work for you, but the code wasn't written with that configuration in mind so there could be cases where you hit weird behaviors like files/directories that reappear even if you deleted them

**Comment by Swivelgames** [comment]
@czheo I would love for this to be a supported use-case, but if it's not, it needs to be an error.

The corruption eventually catches up to you. I tested it exhaustively with several combinations, from lowerdir, upperdir, and neither.

> If the mountpoint is not empty, it results in file contents corruption, instability, crashes, and some bizarre behaviors from different applications and utilities.

It took quite a bit of time for me to notice it at first, but everywhere I deployed this method (other machines, servers, vms, containers) resulted in file corruption _eventually_.

It's a bug either way

1. because it's a supported use case causes file corruption
2. OR, because an unsupported use case that causes file corruption doesn't error when you attempt it.

At minimum, it should warn, and in the case of an interactive terminal, prompt to continue.

The specific use-case is: `mountpoint is not empty`. Whether it's the upperdir or lowerdir is irrelevant, albeit that's when I first discovered this.

I imagine there's quite a bit of silent corruption happening right now given that I've found others using this type of setup.

**Comment by czheo** [comment]
> so if you remove a file in the merged directory

@giuseppe If I use it as purely read-only FS, will it work okay today?

> I would love for this to be a supported use-case, but if it's not, it needs to be an error.

@Swivelgames to be clear, I wasn't saying it shouldn't be an error. I was merely suggesting the MAJOR version should be bumped since it may be a breaking change from user's perspective.

**Comment by giuseppe** [comment]
Yes, that should work
